### PR TITLE
FSDP2/MoE core training stability fixes

### PR DESCRIPTION
**Pull request by ashutoshuiuc**
## Summary
- **Fix NCCL deadlock**: Non-finite grad_norm no longer skips `optimizer.step()` — ALL ranks must call it with FSDP, zero grads before stepping instead
- **Move all-reduce metrics after all micro-batches**: Prevents NCCL deadlock from issuing `dist.all_reduce` while FSDP2 backward gradient reductions are in-flight
- **Patch MoE experts for FSDP2**: Qwen3MoeSparseMoeBlock and Qwen3MoeExperts iterate ALL experts unconditionally for deterministic computation graph with non-reentrant gradient checkpointing
- **Patch non-reentrant checkpoint for MoE**: Tolerant unpack_hook for missing handles from MoE routing differences
- **FSDP2 meta tensor init always enabled**: Handles tied embeddings via `tie_weights()`
- **Batched/coalesced broadcast**: 500MB batches reduce MoE init from minutes to seconds (18K+ params)
- **Guard `output_router_logits`**: Only enable on actual MoE models (`num_local_experts > 0`)

Split from #1298 per maintainer feedback.

Relates to #1297

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1370" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 1]
```diff
@@ -169,31 +169,37 @@ def optimizer_step(
         **kwargs,
     ) -> Optional[Float[torch.Tensor, "1"]]:
         """Perform optimizer step"""
+        import time as _time
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:172]: +1

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 2]
```diff
@@ -169,31 +169,37 @@ def optimizer_step(
...
-                logger.warning(f"grad_norm is not finite: {grad_norm}")
-            optimizer.zero_grad()
-            return grad_norm
+        # NOTE: With FSDP, optimizer.step() involves NCCL collectives. ALL ranks
+        # must call it even if grad_norm is non-finite, otherwise NCCL deadlocks.
+        # We zero_grad before stepping so the non-finite update is harmless.
+        non_finite = grad_norm is not None and not torch.isfinite(grad_norm)
+        if non_finite:
+            logger.warning(f"rank {rank} grad_norm is not finite: {grad_norm}, zeroing grads before step")
+            optimizer.zero_grad(set_to_none=True)
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:196]: Why is this modification needed? We skip `optimizer.step` on all ranks when grad norm is NaN ?

File: skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py [hunk 3]
```diff
@@ -53,6 +55,197 @@
     from skyrl.train.config.config import InferenceEngineConfig
 
 
+def _patch_moe_experts_for_fsdp2(model: nn.Module):
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py:58]: Are these patches still needed with transformers 5.0.0? 

https://github.com/pytorch/pytorch/issues/171355 this seems to be the original issue, and it seems to have been fixed in transformers in February: https://github.com/pytorch/pytorch/issues/171355#issuecomment-3889112286

Comment (SumanthRH) [reply skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py:58]: If so, let's also reorg and move the patches to a separate `patches.py` file

Comment (SumanthRH) [reply skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py:58]: If the patch is needed, let's also verify E2E training with this? A first step would be to ensure some of our GPU CI tests pass (but modified to use Qwen 3 model):

https://github.com/NovaSky-AI/SkyRL/blob/main/tests/backends/skyrl_train/gpu/gpu_ci/test_model_wrapper.py

https://github.com/NovaSky-AI/SkyRL/blob/bcf680f29626bccc725438d902281543e43e7215/tests/backends/skyrl_train/gpu/gpu_ci/test_training_step.py#L2

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 4]
```diff
@@ -169,31 +169,37 @@ def optimizer_step(
...
         if self.max_norm > 0:
+            t0 = _time.time()
             # NOTE (sumanthrh): All `grad_norm`s returned here are the original grad norms before clipping.
             if isinstance(model, FSDP):
                 grad_norm = model.clip_grad_norm_(max_norm=self.max_norm)
             elif isinstance(model, FSDPModule):
                 grad_norm = fsdp2_clip_grad_norm_(model.parameters(), max_norm=self.max_norm)
             else:
                 grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=self.max_norm)
+            logger.info(f"[rank {rank}] clip_grad_norm_ done in {_time.time() - t0:.1f}s, grad_norm={grad_norm}")
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:188]: These seem to be for debugging? Can be removed

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 5]
```diff
@@ -169,31 +169,37 @@ def optimizer_step(
...
+        # We zero_grad before stepping so the non-finite update is harmless.
+        non_finite = grad_norm is not None and not torch.isfinite(grad_norm)
+        if non_finite:
+            logger.warning(f"rank {rank} grad_norm is not finite: {grad_norm}, zeroing grads before step")
+            optimizer.zero_grad(set_to_none=True)
 
+        t0 = _time.time()
         optimizer.step()
-        if scheduler is not None:
+        logger.info(f"[rank {rank}] optimizer.step() done in {_time.time() - t0:.1f}s")
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:200]: again this seems like a debugging log

File: skyrl/backends/skyrl_train/distributed/fsdp_utils.py [hunk 6]
```diff
@@ -293,33 +293,57 @@ def _cast_and_contiguous(tensor, to_contiguous, dtype):
...
+            _flush_batch()
+            batch_tensors.clear()
+            batch_names.clear()
+            batch_bytes = 0
+
+        batch_tensors.append(full_tensor)
+        batch_names.append(param_name)
+        batch_bytes += tensor_bytes
+
+    # Flush remaining tensors
```
Comment (SumanthRH) [inline skyrl/backends/skyrl_train/distributed/fsdp_utils.py:345]: Nice!

**Review by SumanthRH — COMMENTED** [review]
Great overall! Left a few minor comments this time

**Comment by SumanthRH** [comment]
@ashutoshuiuc can you merge from main?

Also I'm curious if you have a stable training curve with MoE models with the FSDP2 backend now. This has been tested before btw, but I'm wondering what model configuration you hit the NCCL deadlock issues and if training is more stable now. Would be good to document the parameter init speedups as well for a 30B model
